### PR TITLE
Add Buddhist calendar calculations

### LINKED ISSUE #1118: Implement the Buddhist calendar
Part of https://github.com/unicode-org/icu4x/issues/534

### PR DESCRIPTION
**Pull request by Manishearth**
This is basically rhe same implementation as the gregorian calendar with a different zero year.

Fixes #1118

### CODE REVIEW

**Review by zbraniecki — DISMISSED** [review]
lgtm. I'd suggest extracting `543` as a const/static.

File: components/calendar/src/buddhist.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+///
+/// [cal]: https://en.wikipedia.org/wiki/Thai_solar_calendar
+pub struct Buddhist;
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+/// The inner date type used for representing Date<Buddhist>
+pub struct BuddhistDateInner(IsoDateInner);
+
+impl Calendar for Buddhist {
+    type DateInner = BuddhistDateInner;
```
Comment (sffc) [inline components/calendar/src/buddhist.rs:31]: Suggestion: just make `DateInner = IsoDateInner`.  Why do we need the wrapper?

File: components/calendar/src/buddhist.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            time: types::Time::try_new(hour, minute, second)?,
+        })
+    }
+}
+
+fn iso_year_as_buddhist(year: IsoYear) -> types::Year {
+    let buddhist_year = year.0 + BUDDHIST_ERA_OFFSET;
+    types::Year {
+        era: types::Era(tinystr8!("be")),
+        number: buddhist_year,
```
Comment (sffc) [inline components/calendar/src/buddhist.rs:136]: Question: this creates negative years in the "be" era.  Is that correct, or is there a BCE-like era in this calendar?

Comment (Manishearth) [reply components/calendar/src/buddhist.rs:136]: It's "correct" in that the calendar is not supposed to be used for past dates, but we already decided that we won't support cutoff dates for calendars and instead will fall back to negative eras.

Comment (sffc) [reply components/calendar/src/buddhist.rs:136]: That's right; I was asking if this calendar system has an era for old dates, something like "BBE" (before buddhist era)?  If not, then negative dates in the current era is fine.

**Review by sffc — APPROVED** [review]

**Comment by Manishearth** [comment]
oops, did not intend to rerequest a review

**Review by sffc — APPROVED** [review]

**Pull request merged**